Match get_attr when compare node

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91657

The pattern can't be matched if one attribute is `_param_constant1` and the other is `_param_constant0`

Large graph:
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant1, ph_0, _tensor_constant0)  {}
```

Pattern graph
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant0, ph_0, _tensor_constant0)  {}
```

Differential Revision: [D42316574](https://our.internmc.facebook.com/intern/diff/D42316574/)

cccclai committed:
Match get_attr when compare node

The pattern can't be matched if one attribute is `_param_constant1` and the other is `_param_constant0`

Large graph:
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant1, ph_0, _tensor_constant0)  {}
```

Pattern graph
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant0, ph_0, _tensor_constant0)  {}
```

Differential Revision: [D42316574](https://our.internmc.facebook.com/intern/diff/D42316574/)

[ghstack-poisoned]

cccclai committed:
Update on "Match get_attr when compare node"


The pattern can't be matched if one attribute is `_param_constant1` and the other is `_param_constant0`

Large graph:
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant1, ph_0, _tensor_constant0)  {}
```

Pattern graph
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant0, ph_0, _tensor_constant0)  {}
```

Differential Revision: [D42316574](https://our.internmc.facebook.com/intern/diff/D42316574/)

[ghstack-poisoned]

cccclai commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -99,6 +100,10 @@ def _nodes_are_equal(self, pn: Node, gn: Node) -> bool:
+            elif pn.op == "get_attr":
+                pn_value = getattr(pn.graph.owning_module, pn.target)
+                gn_value = getattr(gn.graph.owning_module, gn.target)
+                return isinstance(pn_value, torch.Tensor) and isinstance (gn_value, torch.Tensor)
```
I feel like we don't necessarily need to compare tensor data itself, torch.linear(10, 14) and torch.linear(12, 14) will have different weight and they should be matched, thought? 

Also regarding
> legal object
anything else I need to check here?

cccclai committed:
Update on "Match get_attr when compare node"


The pattern can't be matched if one attribute is `_param_constant1` and the other is `_param_constant0`

Large graph:
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant1, ph_0, _tensor_constant0)  {}
```

Pattern graph
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant0, ph_0, _tensor_constant0)  {}
```

Differential Revision: [D42316574](https://our.internmc.facebook.com/intern/diff/D42316574/)

[ghstack-poisoned]

SherlockNoMad commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -91,6 +92,18 @@ def __init__(self, pattern: Graph,
+        gn_value = getattr(gn.graph.owning_module, gn.target)
+        if type(pn_value) != type(gn_value):
+            return False
+        if isinstance(pn_value, torch.Tensor):
```
Please comment on we don't require exact match on tensor values.

SherlockNoMad commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -91,6 +92,18 @@ def __init__(self, pattern: Graph,
+            return False
+        if isinstance(pn_value, torch.Tensor):
+            return isinstance (gn_value, torch.Tensor)
+        else:
```
We should be able handle python literals, int/float/bool ect...

cccclai replied:
Make sense. For python literals, i'm thinking they should be pass if the type is the same maybe?

SherlockNoMad commented:
Can we add some test?

SherlockNoMad reviewed (approved):
Per comment and lint passing.

cccclai committed:
Update on "Match get_attr when compare node"


The pattern can't be matched if one attribute is `_param_constant1` and the other is `_param_constant0`

Large graph:
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant1, ph_0, _tensor_constant0)  {}
```

Pattern graph
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant0, ph_0, _tensor_constant0)  {}
```

Differential Revision: [D42316574](https://our.internmc.facebook.com/intern/diff/D42316574/)

[ghstack-poisoned]

cccclai committed:
Update on "Match get_attr when compare node"


The pattern can't be matched if one attribute is `_param_constant1` and the other is `_param_constant0`

Large graph:
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant1, ph_0, _tensor_constant0)  {}
```

Pattern graph
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant0, ph_0, _tensor_constant0)  {}
```

Differential Revision: [D42316574](https://our.internmc.facebook.com/intern/diff/D42316574/)

[ghstack-poisoned]

cccclai committed:
Update on "Match get_attr when compare node"


The pattern can't be matched if one attribute is `_param_constant1` and the other is `_param_constant0`

Large graph:
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant1, ph_0, _tensor_constant0)  {}
```

Pattern graph
```
        # call_function  addmm_default      aten.addmm.default  (_param_constant0, ph_0, _tensor_constant0)  {}
```

Differential Revision: [D42316574](https://our.internmc.facebook.com/intern/diff/D42316574/)

[ghstack-poisoned]
